### ISSUE TITLE
Publishing with a crates.io mirror has a confusing error message

### ISSUE THREAD

**Issue by ehuss** [rust-lang/cargo#6722]
**Problem**
If one is using a mirror of crates.io, the error message you get when you attempt to publish is confusing.  It says:

```
error: crates cannot be published to crates.io with dependencies sourced from other
registries either publish `bitflags` on crates.io or pull it into this repository
and specify it with a path and version
(crate `bitflags` is pulled from registry `https://github.com/rust-lang/crates.io-index`)
```

The last line clearly says that it is from "crates.io".

**Steps**
1. Check out crates.io index (https://github.com/rust-lang/crates.io-index)
2. Create a project with a dependency from crates.io.
3. Add a `.cargo/config` file to set up the mirror:
```toml
[source.crates-io]
replace-with="mirror"

[source.mirror]
registry = "file:///path/to/crates.io-index"
```
4. `cargo publish --dry-run`

**Possible Solution(s)**
- Display a better error message, possibly instructing the user to disable the mirror.
- Detect that a mirror is being used, and allow it to proceed. This seems like it could be tricky.

**Notes**
- cargo 1.34.0-nightly (5c6aa46e6 2019-02-22)
- Reported by user "calidion" on Discord #crates-io

**Comment by clearloop** [comment]
Same problem.

**Comment by chengyuhui** [comment]
Connection to GitHub is basically trash for some ISPs in China, so a mirror is sometimes required. Having to comment out `~/.cargo/config` every time I need to publish a crate can be annoying.

**Comment by ehuss** [comment]
Closing since I believe this is now fixed via #10907. The new error message is:

```
error: crates-io is replaced with non-remote-registry source registry `mirror`;
include `--registry crates-io` to use crates.io
```
